### PR TITLE
Add no-op methods of public functions

### LINKED ISSUE #12: Provide mechanism to disable Python NVTX ranges
In C/C++, NVTX ranges can be disabled, effectively eliminating any overhead of the nvtx function calls (albeit this overhead should already be small when a tool is not attached).

The Python NVTX ranges should have a similar feature that makes any annotation a no-op.

### PR DESCRIPTION
**Pull request by shwina**
Adds no-op methods of public APIs like `annotate`, which are used if the env var `NVTX_DISABLE` is defined.

Depends on #23 

Closes #12 

Without `NVTX_DISABLE`:

```python
In [1]: from nvtx import annotate

In [2]: @annotate()
   ...: def  foo():
   ...:     pass
   ...:

In [3]: %timeit foo()
328 ns ± 0.411 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

With `NVTX_DISABLE`:

```python
In [1]: from nvtx import annotate

In [2]: @annotate()
   ...: def  foo():
   ...:     pass
   ...:

In [3]: %timeit foo()
49.4 ns ± 0.0631 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

### CODE REVIEW

File: python/nvtx/utils/nop.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    def __call__(self, func):
+        return func
+
+
+def disable(func):
+    # disable the wrapped function
+    @wraps(func)
+    def inner(*args, **kwargs):
+        pass
+    return inner
```
Comment (kkraus14) [inline python/nvtx/utils/nop.py]: This still incurs the overhead of having an inner function call, right? Also, handling `*args, **kwargs` is also surprisingly not free.

What about if we read the environment variable once at library load time and then defined the functions to only `pass` if the environment variable is set?

Comment (shwina) [reply python/nvtx/utils/nop.py]: You're right: using a decorator like this incurs a significant overhead compared with a function that just does `pass`:

```python
In [48]: def foo(x, y):
    ...:     pass
    ...:

In [49]: %timeit foo(1, 2)
121 ns ± 0.789 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

```python
In [51]: @disable
    ...: def foo(x, y):
    ...:     return x + y
    ...:
    ...:

In [52]: %timeit foo(1, 2)
198 ns ± 1.95 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Barring the use of a decorator with an inner function though, our options are:

1. Write duplicate functions with the same signature that `pass`
2.  "Programatically" define the no-op function with `exec`. Something similar was suggested  here: https://stackoverflow.com/a/25460056

Using an [`exec` based solution](https://gist.github.com/shwina/cd515c05d39b2618d1c8563f0507377d) gets us where we want:

```python
In [12]: @disable
    ...: def foo(x, y=1):
    ...:     return x + y
    ...:

In [13]: foo
Out[13]: <function foo(x, y=1)>

In [14]: foo(1, 2)

In [15]: %timeit foo(1, 2)
133 ns ± 0.909 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Comment (shwina) [reply python/nvtx/utils/nop.py]: Discussed offline, we decided  to go with (1) for now and steer clear of `exec`.

In the future, we  might consider a preprocessing step to autogenerate the no-op functions.

File: python/nvtx/nvtx.py [hunk 2]
```diff
@@ -231,3 +233,28 @@ def end_range(range_id):
...
+if not enabled():
+    class annotate(contextlib.nullcontext):
+        def __init__(self, *args, **kwargs):
+            pass
+        def __call__(self, func):
+            return func
+
+    # Could use a decorator here but overheads are significant enough
+    # not to. See https://github.com/NVIDIA/NVTX/pull/24 for discussion.
+    _default = object()
```
Comment (kkraus14) [inline python/nvtx/nvtx.py]: What's the reason for using `_default` instead of `None`?

Comment (shwina) [reply python/nvtx/nvtx.py]: `None` I guess -- I figured there was a reason it was being done in the S/O answer above but `None` is as lightweight as `object()`. Fixed.

File: python/nvtx/nvtx.py [hunk 3]
```diff
@@ -231,3 +233,28 @@ def end_range(range_id):
         The `RangeId` object returned by the `start_range` function.
     """
     libnvtx_end_range(range_id)
+
+
+def enabled():
+    """
+    Returns True if nvtx is enabled.
+    """
+    return not os.getenv("NVTX_DISABLE", False)
```
Comment (kkraus14) [inline python/nvtx/nvtx.py]: If we want this to be part of the public API we likely need to do something like storing the result at runtime in `_enabled` and then return that value in future invocations.

Comment (kkraus14) [reply python/nvtx/nvtx.py]: May also want to rename this to `is_enabled()` just to make it clear that this is a check and not something to call to change a setting.

Comment (shwina) [reply python/nvtx/nvtx.py]: 👍 fixed

**Review by kkraus14 — APPROVED** [review]

**Pull request merged**